src/gui/qgsrichtexteditor.cpp

```diff
@@ -0,0 +1,678 @@
+  dialog.setMinimumWidth( 400 );
+  dialog.setMinimumHeight( 600 );
+  dialog.exec();
+
+  mTextEdit->setHtml( pte->toPlainText() );
+}
+
+void QgsRichTextEditor::clearSource()
+{
+  mTextEdit->clear();
+}
+
+void QgsRichTextEditor::textRemoveFormat()
+{
+  QTextCharFormat fmt;
+  fmt.setFontWeight( QFont::Normal );
+  fmt.setFontUnderline( false );
+  fmt.setFontStrikeOut( false );
+  fmt.setFontItalic( false );
+  fmt.setFontPointSize( 9 );
+//  fmt.setFontFamily     ("Helvetica");
+//  fmt.setFontStyleHint  (QFont::SansSerif);
+//  fmt.setFontFixedPitch (true);
+
+  mActionBold->setChecked( false );
+  mActionUnderline->setChecked( false );
+  mActionItalic->setChecked( false );
+  mActionStrikeOut->setChecked( false );
+  mFontSizeCombo->setCurrentIndex( mFontSizeCombo->findData( 9 ) );
+
+//  QTextBlockFormat bfmt = cursor.blockFormat();
+//  bfmt->setIndent(0);
+
+  fmt.clearBackground();
+
+  mergeFormatOnWordOrSelection( fmt );
+}
+
+void QgsRichTextEditor::textRemoveAllFormat()
+{
+  mActionBold->setChecked( false );
+  mActionUnderline->setChecked( false );
+  mActionItalic->setChecked( false );
+  mActionStrikeOut->setChecked( false );
+  mFontSizeCombo->setCurrentIndex( mFontSizeCombo->findData( 9 ) );
+  QString text = mTextEdit->toPlainText();
+  mTextEdit->setPlainText( text );
+}
+
+void QgsRichTextEditor::textBold()
+{
+  QTextCharFormat fmt;
+  fmt.setFontWeight( mActionBold->isChecked() ? QFont::Bold : QFont::Normal );
+  mergeFormatOnWordOrSelection( fmt );
+}
+
+void QgsRichTextEditor::focusInEvent( QFocusEvent * )
+{
+  mTextEdit->setFocus( Qt::TabFocusReason );
+}
+
+void QgsRichTextEditor::textUnderline()
+{
+  QTextCharFormat fmt;
+  fmt.setFontUnderline( mActionUnderline->isChecked() );
+  mergeFormatOnWordOrSelection( fmt );
+}
+
+void QgsRichTextEditor::textItalic()
+{
+  QTextCharFormat fmt;
+  fmt.setFontItalic( mActionItalic->isChecked() );
+  mergeFormatOnWordOrSelection( fmt );
+}
+
+void QgsRichTextEditor::textStrikeout()
+{
+  QTextCharFormat fmt;
+  fmt.setFontStrikeOut( mActionStrikeOut->isChecked() );
+  mergeFormatOnWordOrSelection( fmt );
+}
+
+void QgsRichTextEditor::textSize( const QString &p )
+{
+  qreal pointSize = p.toDouble();
+  if ( p.toFloat() > 0 )
+  {
+    QTextCharFormat fmt;
+    fmt.setFontPointSize( pointSize );
+    mergeFormatOnWordOrSelection( fmt );
+  }
+}
+
+void QgsRichTextEditor::textLink( bool checked )
+{
+  bool unlink = false;
+  QTextCharFormat fmt;
+  if ( checked )
+  {
+    QString url = mTextEdit->currentCharFormat().anchorHref();
+    bool ok;
+    QString newUrl = QInputDialog::getText( this, tr( "Create a Link" ),
+                                            tr( "Link URL:" ), QLineEdit::Normal,
+                                            url,
+                                            &ok );
+    if ( ok )
+    {
+      fmt.setAnchor( true );
+      fmt.setAnchorHref( newUrl );
+      fmt.setForeground( QApplication::palette().color( QPalette::Link ) );
```

Comment:
This won't take non-default stylesheet styling into account, is this deliberate?